### PR TITLE
feat(condo): DOMA-5993 added incidents export task

### PR DESCRIPTION
**Pull request by Alllex202**

### CODE REVIEW

File: apps/condo/domains/ticket/hooks/useIncidentExportTaskUIInterface.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import { useIntl } from '@open-condo/next/intl'
+
+import { ITask, TASK_REMOVE_STRATEGY } from '@condo/domains/common/components/tasks'
+import { TasksCondoStorage } from '@condo/domains/common/components/tasks/storage/TasksCondoStorage'
+import { TASK_COMPLETED_STATUS } from '@condo/domains/common/constants/tasks'
+import { useDownloadFileFromServer } from '@condo/domains/common/hooks/useDownloadFileFromServer'
+import { IncidentExportTask as IncidentExportTaskApi } from '@condo/domains/ticket/utils/clientSchema'
+
+
+export const useIncidentExportTaskUIInterface = () => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/hooks/useIncidentExportTaskUIInterface.ts:13]: Мb make some abstraction? We have a very similar code in export tickets and contacts

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    fields: {
+
+        status: {
+            schemaDoc: 'Status of the job. Used at client side progress information component',
+            type: 'Select',
+            options: EXPORT_STATUS_VALUES,
+            defaultValue: PROCESSING,
+            isRequired: true,
+            access: {
+                create: canOnlyServerSideWithoutUserRequest,
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:36]: What is the point of `canOnlyServerSideWithoutUserRequest`? Why not `create: false`?

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                update: false,
+            },
+        },
+
+    },
+    hooks: {
+        validateInput: async ({ resolvedData, existingItem, addValidationError }) => {
+            if (existingItem) {
+                if (resolvedData['status'] && existingItem['status'] === COMPLETED) {
+                    addValidationError('status is already completed')
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:173]: please use `GQLError` instead `addValidationError`

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+const { i18n } = require('@open-condo/locales/loader')
+
+const { ERROR, EXCEL } = require('../../common/constants/export')
+const { TASK_WORKER_FINGERPRINT } = require('../../common/constants/tasks')
+const { EXCEL_FILE_META, buildExportFile: buildExportExcelFile } = require('../../common/utils/createExportFile')
+const { findAllByKey } = require('../../common/utils/ecmascript.utils')
+const { getHeadersTranslations, EXPORT_TYPE_INCIDENTS } = require('../../common/utils/exportToExcel')
+const { exportRecordsAsCsvFile, exportRecordsAsXlsxFile } = require('../../common/utils/serverSchema/export')
+const { setLocaleForKeystoneContext } = require('../../common/utils/serverSchema/setLocaleForKeystoneContext')
+const { INCIDENT_STATUS_ACTUAL, INCIDENT_WORK_TYPE_SCHEDULED, INCIDENT_WORK_TYPE_EMERGENCY } = require('../constants/incident')
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:17]: fix imports

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        workFinish: incident.workFinish ? formatDate(incident.workFinish, timeZone) : EMPTY_VALUE,
+        workType: incident.workType ? translations.WorkTypes[incident.workType] : EMPTY_VALUE,
+        organization: incident.organization,
+        textForResident: incident.textForResident || EMPTY_VALUE,
+        createdBy: incident.createdBy,
+        createdAt: incident.createdAt ? formatDate(incident.createdAt, timeZone) : EMPTY_VALUE,
+    }
+}
+
+const buildExportFile = async ({ task, rows }) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:112]: Same as `useIncidentExportTaskUIInterface`.
Should we do something common and not copy the code?

File: apps/condo/domains/ticket/schema/IncidentExportTask.js (hunk 2 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/schema/IncidentExportTask.js:36]: `canOnlyServerSideWithoutUserRequest` is function
```
const canOnlyServerSideWithoutUserRequest = () => false
```
if use simple `false` then you catch error like `Field "exportedRecordsCount" is not defined by type "IncidentExportTaskUpdateInput".`

File: apps/condo/domains/ticket/hooks/useIncidentExportTaskUIInterface.ts (hunk 1 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/hooks/useIncidentExportTaskUIInterface.ts:13]: There is indeed a lot of repetitive code with minimal differences. Moved this logic into a separate hook

File: apps/condo/domains/ticket/tasks/exportIncidents.js (hunk 5 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/tasks/exportIncidents.js:112]: I think in this case it's best to leave it as it is.

Comment (Alllex202) [reply apps/condo/domains/ticket/tasks/exportIncidents.js:112]: There are more places that can change, unlike `useIncidentExportTaskUIInterface `

**Review by nomerdvadcatpyat — APPROVED** [review]

File: apps/condo/domains/common/hooks/useExportTaskUIInterface.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+export const useExportTaskUIInterface = <T extends ExportTaskTypes> ({ clientSchema, schemaName }: UseExportTaskUIInterfaceProps): UseExportTaskUIInterfaceReturn => {
+    const intl = useIntl()
+    const ExportTaskProgressTitle = intl.formatMessage({ id: `tasks.${schemaName}.progress.title` })
+    const ExportTaskProgressDescriptionPreparing = intl.formatMessage({ id: `tasks.${schemaName}.progress.description.preparing` })
+    const ExportTaskProgressDescriptionProcessing = intl.formatMessage({ id: `tasks.${schemaName}.progress.description.processing` })
+    const ExportTaskProgressDescriptionCompleted = intl.formatMessage({ id: `tasks.${schemaName}.progress.description.completed` })
+
+    const { downloadFile } = useDownloadFileFromServer()
+
+    const tryToDownloadFile = async (taskRecord: T) => {
```
Comment (akarjakin) [inline apps/condo/domains/common/hooks/useExportTaskUIInterface.ts:38]: useCallback

File: apps/condo/domains/common/hooks/useExportTaskUIInterface.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        } else {
+            console.error(`File is missing in ${schemaName}`, taskRecord)
+        }
+    }
+
+    /**
+     * We need this separation of behavior from data to determine which behaviour
+     * to use for initial loaded tasks by `__typename` field value
+     */
+    const TaskUIInterface: ITask = {
```
Comment (akarjakin) [inline apps/condo/domains/common/hooks/useExportTaskUIInterface.ts:52]: useMemo

File: apps/condo/domains/ticket/access/IncidentExportTask.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    queryOrganizationEmployeeFor,
+    queryOrganizationEmployeeFromRelatedOrganizationFor,
+} = require('@condo/domains/organization/utils/accessSchema')
+
+
+async function canReadIncidentExportTasks ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (user.isAdmin) return {}
```
Comment (akarjakin) [inline apps/condo/domains/ticket/access/IncidentExportTask.js:21]: what about support users ? staff users ?

File: apps/condo/domains/ticket/access/IncidentExportTask.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    if (user.isAdmin) return true
+
+    if (operation === 'create') {
+        if (get(originalInput, ['user', 'connect', 'id']) !== user.id) return false
+
+        const organizationId = get(originalInput, ['where', 'organization', 'id'])
+        const organizationIdIn = get(originalInput, ['where', 'organization', 'id_in'], [])
+        const organizationIds = uniq(compact([organizationId, ...organizationIdIn]))
+
+        if (organizationIds.length === 0) return false
```
Comment (akarjakin) [inline apps/condo/domains/ticket/access/IncidentExportTask.js:38]: isEmpty(organizationIds)

File: apps/condo/domains/ticket/access/IncidentExportTask.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                    OR: [
+                        queryOrganizationEmployeeFor(user.id),
+                        queryOrganizationEmployeeFromRelatedOrganizationFor(user.id),
+                    ],
+                },
+            ],
+        })
+
+        if (userEmployeeOrganizations.length === organizationIds.length) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/access/IncidentExportTask.js:54]: single row conditional is ok here

File: apps/condo/domains/ticket/access/IncidentExportTask.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+            return true
+        }
+    } else if (operation === 'update') {
+        const task = await getById('IncidentExportTask', itemId)
+
+        if (get(task, 'user') === user.id && get(originalInput, 'status') === CANCELLED) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/access/IncidentExportTask.js:60]: same here

File: apps/condo/domains/ticket/hooks/useIncidentExportToExcelTask.tsx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+export const useIncidentExportToExcelTask = (props: UseIncidentExportToExcelInputType): UseIncidentExportToExcelReturnType => {
+    const intl = useIntl()
+    const ExportAsExcelLabel = intl.formatMessage({ id: 'ExportAsExcel' })
+
+    const { label, where, sortBy, user } = props
+
+    const locale = intl.locale
+    const timeZone = intl.formatters.getDateTimeFormat().resolvedOptions().timeZone
```
Comment (akarjakin) [inline apps/condo/domains/ticket/hooks/useIncidentExportToExcelTask.tsx:40]: useMemo

File: apps/condo/domains/ticket/hooks/useIncidentExportToExcelTask.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        sender: getClientSideSenderInfo(),
+        where,
+        format: EXCEL,
+        sortBy,
+        locale,
+        timeZone,
+        user: { connect: { id: get(user, 'id', null) } },
+    })
+
+    const ExportButton: React.FC<ExportButtonInputType> = useCallback(({ disabled = false, id = 'exportToExcel' }) => (
```
Comment (akarjakin) [inline apps/condo/domains/ticket/hooks/useIncidentExportToExcelTask.tsx:55]: getExportButton

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+                create: true,
+                update: false,
+            },
+        },
+
+    },
+    hooks: {
+        validateInput: async ({ resolvedData, existingItem, context }) => {
+            if (existingItem) {
+                if (resolvedData['status'] && existingItem['status'] === COMPLETED) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:193]: single row conditional is ok here

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+                    throw new GQLError(ERRORS.STATUS_IS_ALREADY_COMPLETED, context)
+                }
+                if (resolvedData['status'] && existingItem['status'] === ERROR) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:196]: same

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+                    throw new GQLError(ERRORS.STATUS_IS_ALREADY_ERROR, context)
+                }
+            }
+        },
+        afterChange: async (args) => {
+            const { updatedItem, operation } = args
+            await setFileMetaAfterChange(args)
+
+            if (operation === 'create') {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:205]: single row conditional is ok here

File: apps/condo/domains/ticket/schema/IncidentExportTask.js [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+                    throw new GQLError(ERRORS.STATUS_IS_ALREADY_COMPLETED, context)
+                }
+                if (resolvedData['status'] && existingItem['status'] === ERROR) {
+                    throw new GQLError(ERRORS.STATUS_IS_ALREADY_ERROR, context)
+                }
+            }
+        },
+        afterChange: async (args) => {
+            const { updatedItem, operation } = args
+            await setFileMetaAfterChange(args)
```
Comment (akarjakin) [inline apps/condo/domains/ticket/schema/IncidentExportTask.js:203]: empty line

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    Actual: i18n('incident.status.actual', { locale }),
+    NotActual: i18n('incident.status.notActual', { locale }),
+    WorkTypes: {
+        [INCIDENT_WORK_TYPE_SCHEDULED]: i18n('incident.workType.scheduled', { locale }),
+        [INCIDENT_WORK_TYPE_EMERGENCY]: i18n('incident.workType.emergency', { locale }),
+    },
+})
+
+const classifiersToString = (classifiers = []) => {
+    if (classifiers.length < 1) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:44]: single row conditional is ok here

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+        return EMPTY_VALUE
+    }
+
+    const categories = uniq(classifiers
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:48]: please use `compact` too

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+        .map(({ category }) => category))
+        .join(', ')
+    const problems = uniq(classifiers
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:51]: same

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+
+    const incidentProperties = await loadIncidentPropertiesForExcelExport({
+        where: {
+            incident: { id: incident.id },
+            deletedAt: null,
+        },
+    })
+
+    const incidentClassifierIncidents = await loadIncidentClassifierIncidentsForExcelExport({
+        where: {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:75]: content is same here, can be moved to a variable and reused

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+
+    const classifierIds = uniq(incidentClassifierIncidents.map(({ classifier }) => classifier))
+
+    const classifiers = await loadIncidentClassifiersForExcelExport({
+        where: {
+            id_in: classifierIds,
+        },
+    })
+
+    const properties = incidentProperties
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:89]: single line please

**Review by ekabardinsky — APPROVED** [review]

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+        meta: {
+            listkey: 'IncidentExportTask',
+            id,
+        },
+    }
+}
+
+async function exportIncidents (taskId) {
+    if (!taskId) {
+        taskLogger.error({
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:147]: single line is ok here

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+            msg: 'taskId is undefined',
+        })
+        throw new Error('taskId is undefined')
+    }
+
+    const { keystone: context } = await getSchemaCtx('IncidentExportTask')
+
+    const task = await IncidentExportTask.getOne(context, { id: taskId })
+    if (!task) {
+        taskLogger.error({
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:157]: single line is ok here

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+            msg: `No task with id "${taskId}"`,
+        })
+        throw new Error(`No task with id "${taskId}"`)
+    }
+
+    const { where, sortBy, format, locale } = task
+
+    try {
+        if (!locale) {
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:166]: single line is ok here

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+
+        const convertRecordToFileRow = async (incident) => await incidentToRow({
+            task,
+            incident,
+            translations,
+        })
+
+        const loadRecordsBatch = async (offset, limit) => {
+            const incidents = await incidentLoader.loadChunk(offset, limit)
+            this.progress(Math.floor(offset / totalRecordsCount * 100))
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:186]: empty line

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+            return incidents
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:187]: and here

File: apps/condo/domains/ticket/tasks/exportIncidents.js [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+            ...BASE_ATTRIBUTES,
+            status: ERROR,
+        })
+
+        taskLogger.error({
+            msg: 'Failed to export incidents',
+            data: { id: taskId },
+            err,
+        })
+        throw err
```
Comment (akarjakin) [inline apps/condo/domains/ticket/tasks/exportIncidents.js:227]: empty line

File: apps/condo/domains/ticket/utils/testSchema/index.js [hunk 29]
```diff
@@ -853,6 +840,40 @@ async function updateTestUserFavoriteTicket (client, id, extraAttrs = {}) {
...
+    if (!user || !user.id) throw new Error('no user.id')
+    const sender = { dv: 1, fingerprint: faker.random.alphaNumeric(8) }
+
+    const attrs = {
+        dv: 1,
+        sender,
+        format: EXCEL,
+        where: {},
+        sortBy: ['createdAt_DESC'],
+        locale: 'ru',
```
Comment (akarjakin) [inline apps/condo/domains/ticket/utils/testSchema/index.js:854]: better to randomize it

File: apps/condo/migrations/20230424210801-0263_incidentexporttaskhistoryrecord_incidentexporttask.js [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+ALTER TABLE "IncidentExportTask" ADD CONSTRAINT "IncidentExportTask_createdBy_0a6be750_fk_User_id" FOREIGN KEY ("createdBy") REFERENCES "User" ("id") DEFERRABLE INITIALLY DEFERRED;
+ALTER TABLE "IncidentExportTask" ADD CONSTRAINT "IncidentExportTask_updatedBy_27aa8f50_fk_User_id" FOREIGN KEY ("updatedBy") REFERENCES "User" ("id") DEFERRABLE INITIALLY DEFERRED;
+ALTER TABLE "IncidentExportTask" ADD CONSTRAINT "IncidentExportTask_user_0ccaac00_fk_User_id" FOREIGN KEY ("user") REFERENCES "User" ("id") DEFERRABLE INITIALLY DEFERRED;
+CREATE INDEX "IncidentExportTask_createdAt_bced5491" ON "IncidentExportTask" ("createdAt");
+CREATE INDEX "IncidentExportTask_updatedAt_6d9d1118" ON "IncidentExportTask" ("updatedAt");
+CREATE INDEX "IncidentExportTask_deletedAt_9ed50375" ON "IncidentExportTask" ("deletedAt");
+CREATE INDEX "IncidentExportTask_createdBy_0a6be750" ON "IncidentExportTask" ("createdBy");
+CREATE INDEX "IncidentExportTask_updatedBy_27aa8f50" ON "IncidentExportTask" ("updatedBy");
+CREATE INDEX "IncidentExportTask_user_0ccaac00" ON "IncidentExportTask" ("user");
+COMMIT;
```
Comment (akarjakin) [inline apps/condo/migrations/20230424210801-0263_incidentexporttaskhistoryrecord_incidentexporttask.js:25]: please make sure you've added all indexes on all fields that could be used to sort/filter data

File: apps/condo/domains/ticket/access/IncidentExportTask.js (hunk 8 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/access/IncidentExportTask.js:21]: supports don't have any privileges

**Pull request merged**